include/sparrow/null_layout.hpp

```diff
@@ -0,0 +1,256 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or mplied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <cstddef>
+#include <functional>
+#include <optional>
+#include <ranges>
+
+#include "sparrow/data_type.hpp"
+#include "sparrow/iterator.hpp"
+
+namespace sparrow
+{
+
+    /*
+     * @class empty_iterator
+     *
+     * @brief Iterator used by the null_layout class.
+     *
+     * @tparam T the value_type of the iterator
+     */
+    template <class T>
+    class empty_iterator : public iterator_base<
+                               empty_iterator<T>,
+                               T,
+                               std::contiguous_iterator_tag,
+                               T>
+    {
+    public:
+
+        using self_type = empty_iterator<T>;
+        using base_type = iterator_base<
+            self_type,
+            T,
+            std::contiguous_iterator_tag,
+            T>;
+        using reference = typename base_type::reference;
+        using difference_type = typename base_type::difference_type;
+
+        explicit empty_iterator(difference_type index = difference_type()) noexcept;
+
+    private:
+
+        reference dereference() const;
+        void increment();
+        void decrement();
+        void advance(difference_type n);
+        difference_type distance_to(const self_type& rhs) const;
+        bool equal(const self_type& rhs) const;
+        bool less_than(const self_type& rhs) const;
+
+        difference_type m_index;
+
+        friend class iterator_access;
+    };
+
+    /*
+     * @class null_layout
+     *
+     * @brief Memory-efficient layout for the Null data type.
+     *
+     * This layout is a memory-efficient layout for the Null data type where
+     * all values are null. In this case, no memory buffers are allocated.
+     */
+    class null_layout
+    {
+    public:
+
+        using inner_value_type = null_type;
+        using value_type = std::optional<inner_value_type>;
+        using iterator = empty_iterator<value_type>;
+        using const_iterator = empty_iterator<value_type>;
+        using reference = iterator::reference;
+        using const_reference = const_iterator::reference;
+        using size_type = std::size_t;
+        using difference_type = iterator::difference_type;
+        using iterator_tag = std::contiguous_iterator_tag;
+
+
+        using const_value_iterator = empty_iterator<int>;
+        using const_bitmap_iterator = empty_iterator<bool>;
+
+        using const_value_range = std::ranges::subrange<const_value_iterator>;
+        using const_bitmap_range = std::ranges::subrange<const_bitmap_iterator>;
+
+        explicit null_layout(array_data& data);
+        void rebind_data(array_data& data);
+
+        size_type size() const;
```

Comment:
What about adding empty() ?

Reply:
We don't implement it for other layout classes. These layout classes are not meant to be used by the end-user, but are meant to have the minimal reqiured API to implement the `typed_array` class, I think implementing empty in this latter one is enough.